### PR TITLE
Display the roles of the logged-in user in the Web Debug Toolbar

### LINKED ISSUE #42763: Show user roles in web developer toolbar security widget
**Description**
Include current user roles in toolbar security widget. They can be viewed by clicking the widget, but it's an extra page load. If roles were included, they would probably be the most frequently changing aspect, which also points to them being one of the most useful, especially for role assignment and user access development.

**Example**
image

### PR DESCRIPTION
**Pull request by NicoHaase**
| Q             | A
| ------------- | ---
| Branch?       | 5.4 for features
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | yes/no 
| Tickets       | Fix #42763
| License       | MIT
| Doc PR        | -

This PR adds the roles of the logged-in user to the WDT

### CODE REVIEW

File: src/Symfony/Bundle/SecurityBundle/Resources/views/Collector/security.html.twig [hunk 1]
```diff
@@ -41,6 +41,11 @@
                             <span class="sf-toolbar-status sf-toolbar-status-{{ is_authenticated ? 'green' : 'yellow' }}">{{ is_authenticated ? 'Yes' : 'No' }}</span>
                         </div>
 
+                        <div class="sf-toolbar-info-piece">
+                            <b>Roles</b>
+                            <span>{{ collector.roles|join(', ') }}</span>
```
Comment (derrabus) [inline src/Symfony/Bundle/SecurityBundle/Resources/views/Collector/security.html.twig]: Should we limit the output here? I'm afraid we break the toolbar for apps with many roles. 🙈

Comment (ro0NL) [reply src/Symfony/Bundle/SecurityBundle/Resources/views/Collector/security.html.twig]: was checking :grimacing: i have 200+ roles :}

Comment (derrabus) [reply src/Symfony/Bundle/SecurityBundle/Resources/views/Collector/security.html.twig]: 😱

Comment (ro0NL) [reply src/Symfony/Bundle/SecurityBundle/Resources/views/Collector/security.html.twig]: yes. It's a symptom of permissions-as-roles. In that sense ive no idea what a reasonable no. of roles is, nor about an average role name length. I do agree the current toolbar has limited UI to display many data.

Comment (NicoHaase) [reply src/Symfony/Bundle/SecurityBundle/Resources/views/Collector/security.html.twig]: Yeah, good point - limiting would be a good idea. Do you have any preference about this. I would see the following options:

- showing the first m roles, and something like "and n more"
- having something like a modal that could be opened if there are more than m roles

Comment (ro0NL) [reply src/Symfony/Bundle/SecurityBundle/Resources/views/Collector/security.html.twig]: Perhaps `N roles(s)` is sufficient?  With eg. a title attribute for the actual role names ... 

(ignore my 200+ roles, it's broken design)

Comment (NicoHaase) [reply src/Symfony/Bundle/SecurityBundle/Resources/views/Collector/security.html.twig]: IMO: If this should be a good addition, as requested in #42763, this should not only contain the number of assigned roles

Comment (derrabus) [reply src/Symfony/Bundle/SecurityBundle/Resources/views/Collector/security.html.twig]: We should agree on a threshold above which we just display the number of roles. If a user has 20 roles and we only display three arbitrary ones, I don't think the interface is helpful. If the number of roles is below the threshold, we simply list them.

Comment (NicoHaase) [reply src/Symfony/Bundle/SecurityBundle/Resources/views/Collector/security.html.twig]: @derrabus do you have any specific number in mind?

Comment (derrabus) [reply src/Symfony/Bundle/SecurityBundle/Resources/views/Collector/security.html.twig]: Dunno, five? 🤷🏻‍♂️

Comment (fd6130) [reply src/Symfony/Bundle/SecurityBundle/Resources/views/Collector/security.html.twig]: I think one should be enough, if they wish to know more detail about current roles just tab in the profiler and take a look.

Hard to imagine how does it look like when rendering lots of roles in a small UI.

Comment (ro0NL) [reply src/Symfony/Bundle/SecurityBundle/Resources/views/Collector/security.html.twig]: > this should not only contain the number of assigned roles

hence i suggested to use a title attribute and avoid this discussion :)

Comment (wouterj) [reply src/Symfony/Bundle/SecurityBundle/Resources/views/Collector/security.html.twig]: I like @ro0NL's suggestion of using the title attribute:

image

I think 1 role is a good threshold value :smile: . In this example screenshot, the user only has 4 roles and it already wouldn't fit the normal modal.
Also, I always recommend people to limit the number of roles a user has as close to 1 as possible. Maybe complex applications need 3+ roles, but generally you should be able to manage 1 or 2 roles if you split roles from permissions.

Comment (ro0NL) [reply src/Symfony/Bundle/SecurityBundle/Resources/views/Collector/security.html.twig]: i agree the dashed underline is likely enough to trigger a user hover, and go from there.

i'd opt out from a random number to display roles for

we could spare out "Logged in as" in favor of the "user icon" as in the toolbar itself, if it's not redundant already :}

we could spare out "Authenticated: yes" in favor of the negative edge case (eg. "Not authenticated" as a single badge)

"Action" looks redundant :) could be some "logout icon" too

last but not least, IIUC the title attribute is limited to 512 chars. Perhaps we should invent some CSS only tooltip util.

Comment (wouterj) [reply src/Symfony/Bundle/SecurityBundle/Resources/views/Collector/security.html.twig]: Please keep this discussion focussed and discuss cleaning up other things in another issue/pr :)

I'm 100% in favor of keeping this as simple as possible. Let's not increase difficulty with CSS/js tooltips. The longest role I've seen is Symfony's own `ROLE_PREVIOUS_ADMIN`, which is 19 characters long (meaning a user can have 24 roles before the title attribute overflows).

Comment (ro0NL) [reply src/Symfony/Bundle/SecurityBundle/Resources/views/Collector/security.html.twig]: I never implied "JS". We have a hard limit on title attributes.

edit: the 512 limit is IE only it seems. Max-width is default on title attributes ... let's do that :+1:

**Comment by NicoHaase** [comment]
So, I'm having a new shot after all that feedback - thank you guys for commenting!

Bildschirmfoto von 2021-09-06 07-59-54

**Comment by derrabus** [comment]
@ro0NL Can you test this PR with your 200+ roles app? 🙃

**Review by wouterj — APPROVED** [review]
Looks great :)

**Review by derrabus — APPROVED** [review]

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @NicoHaase.

**Pull request merged**